dircond.1
.\" dircond - directory content watcher daemon -*- nroff -*-
.\" Copyright (C) 2012, 2013 Sergey Poznyakoff
.\"
.\" Dircond is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 3 of the License, or (at your
.\" option) any later version.
.\"
.\" Dircond is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with dircond. If not, see <http://www.gnu.org/licenses/>.
.TH DIRCOND 1 "March 8, 2013" "DIRCOND" "Dircond User Reference"
.SH NAME
dircond \- directory content watcher
.SH SYNOPSIS
\fBdircond\fR [\fB\-dft\fR] [\fB\-F\fR \fIFACILITY\fR] [\fB\-L\fR \fITAG\fR]\
 [\fB\-P\fR \fIFILE\fR]\
 [\fB\-u\fR \fIUSER\fR]\
 [\fICONFIG\fR]

.B dircond -h

.B dircond -V
.SH DESCRIPTION
.B Dircond
monitors a set of directories on the file system and reacts when
a file system event occurs in any of them.  Directories and events
to monitor are specified in the configuration file.  When an event
occurs the utility reacts by invoking an external command configured
for that event.
.PP
The following events can be monitored by the program:
.TP
.B access
A file was accessed.
.TP
.B attrib
A file's metadata changed.
.TP
.B close_write
A writable file was closed.
.TP
.B close_nowrite
An unwritable file closed.
.TP
.B create
A file was created.
.TP
.B delete
A file was deleted.
.TP
.B modify
A file was modified.
.TP
.B moved_from
A file was moved into a monitored directory.
.TP
.B moved_to
A file was moved out from a monitored directory.
.TP
.B open
A file was opened.
.PP
When an event has been detected,
.B dircond
invokes the handler configured for it.  The handler is invoked in the
directory where the event occured.  Additional information about the
event is passed in the following environment variables:
.TP
.B DIRCOND_EVENT
Event symbolic name (see the event table above). 
.TP
.B DIRCOND_EVENT_CODE
Event code as obtained from
.BR inotify (8).
.TP
.B DIRCOND_FILE
The name of the affected file.
.PP
Apart from these variables, the
handler environment is inherited from the calling process. If you need
to restrict the environment, start
.B dircond
via
.BR env (1).
.SH OPTIONS
.TP
.B \-d
Increase debug verbosity level.
.TP
.B \-f
Run in the foreground.
.TP
\fB\-F\fR \fIFACILITY\fR
Log under this syslog facility.  Allowed values for \fIFACILITY\fR are
a decimal number or any of the following symbolic names:
.BR auth ,
.BR authpriv ,
.BR cron ,
.BR daemon ,
.BR ftp ,
.BR local0 " through " local7 ,
.BR lpr ,
.BR mail ,
.BR news ,
.BR user ,
and
.BR uucp .

The option \fB\-F 0\fR directs logging to the standard error.
.TP
\fB\-L\fR \fITAG\fR
Log with this syslog tag.
.TP
\fB\-P\fR \fIFILE\fR
Write PID to \fIFILE\fR.
.TP
\fB\-t\fR
Check configuration file for errors and exit.
.TP
\fB\-u\fR \fIUSER\fR
Run as this \fIUSER\fR.
.TP
\fB\-h\fR
Output a terse help summary and exit.
.TP
\fB\-V\fR
Print program version and copyright information.
.SH CONFIGURATION
The default configuration file is
.BR /etc/dircond.conf .
If a file name is supplied as an argument to the program, that file
will be read instead.
.PP
The configuration file has the traditional line-oriented format.  Very
long lines can be split into several physical lines by ending each of
them with a backslash immediately followed by a newline character.
.PP
Comments are introduced with a sharp sign and extend to the end of
logical line.
.PP
Empty lines are ignored.
.SS Global settings
.TP
\fBforeground\fR [\fBon\fR|\fBoff\fR]
Remain in foreground (if \fBon\fR or no argument given), or detach
from the controlling terminal and become a daemon (if \fBoff\fR).
The latter is the default.  See also the \fB\-f\fR option.
.TP
\fBdebug\fR \fIN\fR
Set debugging level.  See also the \fB\-d\fR option.
.TP
\fBpidfile\fR \fIFILE\fR
Store PID of the started process in \fIFILE\fR.  See also the
\fB\-P\fR option.
.TP
\fBuser\fR \fINAME\fR
Run as user \fINAME\fR.  See also the \fB\-u\fR option.
.SS Logging
The \fBsyslog\fR statement configures where the log output goes.  By
default, the program logs to standard error until it finishes parsing
the configuration file and processing command line options.  From then
on it begins logging to syslog with the facility \fBdaemon\fR.  The
syslog settings can be altered using this statement.
.TP
.B syslog off
Disable syslog output, log everything to standard error instead.  This
setting has effect only when running in the foreground.
.TP
\fBsyslog facility\fR \fIFNAME\fR
Log to syslog facility \fIFNAME\fR.  The latter can be either 
a decimal number or any of the following symbolic names:
.BR auth ,
.BR authpriv ,
.BR cron ,
.BR daemon ,
.BR ftp ,
.BR local0 " through " local7 ,
.BR lpr ,
.BR mail ,
.BR news ,
.BR user ,
and
.BR uucp .
.TP
\fBsyslog tag\fR \fISTRING\fR
Log with the given syslog tag.
.SS Handler options
Handler options affect all event handlers defined in the \fBonevent\fR
statements appearing below them.
.TP 
\fBoption capture\fR \fIstream\fR [\fIenable\fR]
Capture the output stream of a handler.  The \fIstream\fR parameter
specifies which stream to capture.  Its possible values are:
\fBstdout\fR for the standard output, \fBstderr\fR for the standard
error, or \fBboth\fR for both streams.  Optional \fIenable\fR
parameter can be \fBon\fR to enable capturing and \fBoff\fR to disable
it.

When logging to syslog, the data read from the captured
.B stdout
will be logged with the priority
.BR LOG_INFO ,
and
.B stderr
will be logged with the priority
.BR LOG_ERROR .
.TP
\fBoption wait\fR
Wait for the handler to terminate before handling subsequent events.
This is the default.
.TP
\fBoption nowait\fR
Do not wait for the handler to terminate.  Start it in the background
and continue processing requests.
.TP
\fBoption timeout\fR \fIN\fR
Sets timeout in seconds for handler runtime.  The default value is 5
seconds.  Note, that the timeout setting affects both \fBwait\fR and
\fBnowait\fR handlers, i.e. any handler that runs longer
than \fIN\fR seconds will be terminated by
.B SIGKILL.
.SS Event handling
.TP
\fBevent\fR \fINAME\fR \fIEVENT-ID\fR [\fIEVENT-ID\fR...]
Add listed events to the event set \fINAME\fR, creating it if it does
not exist.  Each \fIEVENT-ID\fR is either a predefined event name (see
the \fBDESCRIPTION\fR section) or a \fINAME\fR of an event 
set defined in a previous
.B event
statement.
.TP
\fBpath\fR \fINAME\fR \fIDIR\fR [\fBrecursive\fR [\fIN\fR]]
Add directory \fIDIR\fR to the pathset \fINAME\fR.  If \fBrecursive\fR
is specified, all directories below \fIDIR\fR are also added to the
pathset.  If \fBrecursive\fR is followed by a decimal number \fIN\fR,
recursive inclusion is limited to directories located no deeper than
\fIN\fR levels of nesting.
.TP
\fBonevent\fR \fIEVENT-ID\fR \fBin\fR \fIPATH-NAME\fR \fBrun\fR \fICOMMAND\fR\
 [\fBas\fR \fIUSER\fR]
Configures the handler for events occurring in the given pathset.
\fICOMMAND\fR is the full pathname of an external program.  This
program will be started when one of the events from \fIEVENT-ID\fR
occur for any file in directories comprising the pathset \fIPATH-NAME\fR.
If the optional \fBas\fR \fIUSER\fR clause is supplied, the program
will be run with the UID and (main and supplementary) GIDs of \fIUSER\fR.
The latter must be either a valid user name or a numeric UID of an
existing user.
.SH EXAMPLE
.sp
.nf
.in +2
# Store the PID of the master process in this file:
pidfile /var/run/dircond.pid

# Log using syslog facility \fBlocal0\fR.
syslog facility local0

# Define event set \fBupdate\fR to match any modifications
# of file contents.
event update modify close_write moved_to

# Define event set \fBremove\fR to cover removal of a file
# from the watched directory.
event remove delete moved_from

# Add \fB/etc\fR and its subdirectories to pathset \fBconfigs\fR.
path configs /etc recursive

# Don't allow event handler to run longer than 10 seconds.
option timeout 10
# Capture both standard output and error and divert them
# to the log file.
option capture both

# If any file gets updated in \fB/etc\fR or any of its
# subdirectories, call the program \fB/usr/bin/confchange\fR.
onevent update in configs run /usr/bin/confchange as daemon

# Another handler, for removals in the configuration directory.
onevent delete in configs run /usr/bin/confdel
.in
.fi
.SH "EXIT CODE"
.IP 0
Successful termination.
.IP 1
Command line usage error.
.IP 2
Another error occurred.
.SH "SEE ALSO"
.BR inotify (8).
.SH AUTHORS
Sergey Poznyakoff
.SH "BUG REPORTS"
Report bugs to <gray+dircond@gnu.org.ua>.
.SH COPYRIGHT
Copyright \(co 2012, 2013 Sergey Poznyakoff
.br
.na
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
.ad
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.\" Local variables:
.\" eval: (add-hook 'write-file-hooks 'time-stamp)
.\" time-stamp-start: ".TH [A-Z_][A-Z0-9_.\\-]* [0-9] \""
.\" time-stamp-format: "%:B %:d, %:y"
.\" time-stamp-end: "\""
.\" time-stamp-line-limit: 20
.\" end:



NEWS
dircond -- history of user-visible changes. 2013-06-01
Copyright (C) 2012, 2013 Sergey Poznyakoff
See the end of file for copying conditions.

Please send dircond bug reports to <gray+dircond@gnu.org.ua>

Version 3.0, 2013-06-01

Introduce configuration file.


Version 2.0, 2013-01-11

* Change command line handling

Dircond accepts handler definitions interspersed with directory
name.  This allows for having different per-directory handlers.

Similarly, the autowatch options ("-a" and "-l") affect the
directories specified after them in the command line and their effect
is changed by the next occurrence of the corresponding option.

* Bugfixes

Fix a minor error in fd_set handling.


Version 1.0, 2012-12-30

Initial release


=========================================================================
Copyright information:

Copyright (C) 2012, 2013 Sergey Poznyakoff

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.

Local variables:
mode: outline
paragraph-separate: "[  ]*$"
eval: (add-hook 'write-file-hooks 'time-stamp)
time-stamp-start: "changes. "
time-stamp-format: "%:y-%02m-%02d"
time-stamp-end: "\n"
end:




src/dircond.c
/* dircond - directory content watcher daemon
   Copyright (C) 2012, 2013 Sergey Poznyakoff

   Dircond is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 3 of the License, or (at your
   option) any later version.

   Dircond is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with dircond. If not, see <http://www.gnu.org/licenses/>. */

#include "dircond.h"
#include <stdarg.h>
#include <fcntl.h>
#include <getopt.h>
#include <pwd.h>
#include <grp.h>
#include <signal.h>
#include <grecs.h>

#ifndef SYSCONFDIR
# define SYSCONFDIR "/etc"
#endif
#define DEFAULT_CONFFILE SYSCONFDIR "/dircond.conf"

/* Configuration settings */
const char *program_name;         /* This program name */
const char *conffile = DEFAULT_CONFFILE;
int foreground;                   /* Remain in the foreground */
char *tag;                        /* Syslog tag */
int facility = -1;                /* Use this syslog facility for logging.
				     -1 means log to stderr */
int syslog_include_prio;
int debug_level;                  /* Debug verbosity level */
char *pidfile = NULL;             /* Store PID to this file */
char *user = NULL;                /* User to run as */

int log_to_stderr = LOG_DEBUG;


/* Diagnostic functions */
const char *
severity(int prio)
{
	switch (prio) {
	case LOG_EMERG:
		return "EMERG";
	case LOG_ALERT:
		return "ALERT";
	case LOG_CRIT:
		return "CRIT";
	case LOG_ERR:
		return "ERROR";
	case LOG_WARNING:
		return "WARNING";
	case LOG_NOTICE:
		return "NOTICE";
	case LOG_INFO:
		return "INFO";
	case LOG_DEBUG:
		return "DEBUG";
	}
	return NULL;
}

void
vdiag(int prio, const char *fmt, va_list ap)
{
	const char *s;
	va_list tmp;
	
	if (log_to_stderr >= prio) {
		fprintf(stderr, "%s: ", program_name);
		s = severity(prio);
		if (s)
			fprintf(stderr, "[%s] ", s);
		va_copy(tmp, ap);
		vfprintf(stderr, fmt, tmp);
		fputc('\n', stderr);
		va_end(tmp);
	}

	if (facility > 0) {
		if (syslog_include_prio && (s = severity(prio)) != NULL) {
			static char *fmtbuf;
			static size_t fmtsize;
			size_t len = strlen(fmt) + strlen(s) + 4;
			char *p;
			
			if (len > fmtsize) {
				fmtbuf = erealloc(fmtbuf, len);
				fmtsize = len;
			}

			p = fmtbuf;
			*p++ = '[';
			while (*s)
				*p++ = *s++;
			*p++ = ']';
			*p++ = ' ';
			while (*p++ = *fmt++);
			vsyslog(prio, fmtbuf, ap);
		} else			
			vsyslog(prio, fmt, ap);
	}
}

void
diag(int prio, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vdiag(prio, fmt, ap);
	va_end(ap);
}
	
void
debugprt(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vdiag(LOG_DEBUG, fmt, ap);
	va_end(ap);
}

/* Memory allocation with error checking */
void *
emalloc(size_t size)
{
	void *p = malloc(size);
	if (!p) {
		diag(LOG_CRIT, "not enough memory");
		exit(2);
	}
	return p;
}

void *
ecalloc(size_t nmemb, size_t size)
{
	void *p = calloc(nmemb, size);
	if (!p) {
		diag(LOG_CRIT, "not enough memory");
		exit(2);
	}
	return p;
}

void *
erealloc(void *ptr, size_t size)
{
	void *p = realloc(ptr, size);
	if (!p) {
		diag(LOG_CRIT, "not enough memory");
		exit(2);
	}
	return p;
}

char *
estrdup(const char *str)
{
	size_t len = strlen(str);
	char *p = emalloc(len + 1);
	memcpy(p, str, len);
	p[len] = 0;
	return p;
}

/* Create a full file name from directory and file name */
char *
mkfilename(const char *dir, const char *file)
{
	char *tmp;
	size_t dirlen = strlen(dir);
	size_t fillen = strlen(file);
	size_t len;

	while (dirlen > 0 && dir[dirlen-1] == '/')
		dirlen--;

	len = dirlen + (dir[0] ? 1 : 0) + fillen;
	tmp = malloc(len + 1);
	if (tmp) {
		memcpy(tmp, dir, dirlen);
		if (dir[0])
			tmp[dirlen++] = '/';
		memcpy(tmp + dirlen, file, fillen);
		tmp[len] = 0;
	}
	return tmp;
}

int
trans_strtotok(struct transtab *tab, const char *str, int *ret)
{
	for (; tab->name; tab++)
		if (strcmp(tab->name, str) == 0) {
			*ret = tab->tok;
			return 0;
		}
	return -1;
}

char *
trans_toktostr(struct transtab *tab, int tok)
{
	for (; tab->name; tab++)
		if (tab->tok == tok)
			return tab->name;
	return NULL;
}

char *
trans_toknext(struct transtab *tab, int tok, int *next)
{
	int i;
	
	for (i = *next; tab[i].name; i++)
		if (tab[i].tok & tok) {
			*next = i + 1;
			return tab[i].name;
		}
	*next = i;
	return NULL;
}

char *
trans_tokfirst(struct transtab *tab, int tok, int *next)
{
	*next = 0;
	return trans_toknext(tab, tok, next);
}


/* Command line processing and auxiliary functions */

static void
set_program_name(const char *arg)
{
	char *p = strrchr(arg, '/');
	if (p)
		program_name = p + 1;
	else
		program_name = arg;
}


void
signal_setup(void (*sf) (int))
{
	signal(SIGTERM, sf);
	signal(SIGQUIT, sf);
	signal(SIGINT, sf);
	signal(SIGHUP, sf);
	signal(SIGALRM, sf);
	signal(SIGUSR1, sf);
	signal(SIGUSR2, sf);
	signal(SIGCHLD, sf);
}

void
storepid(const char *pidfile)
{
	FILE *fp = fopen(pidfile, "w");
	if (!fp) {
		diag(LOG_ERR, "cannot open pidfile %s for writing: %s",
		     pidfile, strerror(errno));
	} else {
		fprintf(fp, "%lu\n", (unsigned long) getpid());
		fclose(fp);
	}
}

static int
membergid(gid_t gid, size_t gc, gid_t *gv)
{
	int i;
	for (i = 0; i < gc; i++)
		if (gv[i] == gid)
			return 1;
	return 0;
}

static void
get_user_groups(uid_t uid, size_t *pgidc, gid_t **pgidv)
{
	size_t gidc = 0, n = 0;
	gid_t *gidv = NULL;
	struct passwd *pw;
	struct group *gr;

	pw = getpwuid(uid);
	if (!pw) {
		diag(LOG_ERR, 0, "no used with UID %lu",
		     (unsigned long)uid);
		exit(2);
	}
	
	n = 32;
	gidv = ecalloc(n, sizeof(gidv[0]));
		
	gidv[0] = pw->pw_gid;
	gidc = 1;
	
	setgrent();
	while (gr = getgrent()) {
		char **p;
		for (p = gr->gr_mem; *p; p++)
			if (strcmp(*p, pw->pw_name) == 0) {
				if (n == gidc) {
					n += 32;
					gidv = erealloc(gidv,
							n * sizeof(gidv[0]));
				}
				if (!membergid(gr->gr_gid, gidc, gidv))
					gidv[gidc++] = gr->gr_gid;
			}
	}
	endgrent();
	*pgidc = gidc;
	*pgidv = gidv;
}

void
setuser(const char *user)
{
	struct passwd *pw;
	size_t gidc;
	gid_t *gidv;
		
	pw = getpwnam(user);
	if (!pw) {
		diag(LOG_CRIT, "getpwnam(%s): %s", user, strerror(errno));
		exit(2);
	}
	if (pw->pw_uid == 0)
		return;

	get_user_groups(pw->pw_uid, &gidc, &gidv);
	if (setgroups(gidc, gidv) < 0) {
		diag(LOG_CRIT, "setgroups: %s", strerror(errno));
		exit(2);
	}
	free(gidv);

	if (setgid(pw->pw_gid)) {
		diag(LOG_CRIT, "setgid(%lu): %s", (unsigned long) pw->pw_gid,
		     strerror(errno));
		exit(2);
	}
	if (setuid(pw->pw_uid)) {
		diag(LOG_CRIT, "setuid(%lu): %s", (unsigned long) pw->pw_uid,
		     strerror(errno));
		exit(2);
	}
}

void
ev_log(int flags, struct dirwatcher *dp)
{
	int i;
	char *p;
	
	if (debug_level > 0) {
		for (p = trans_tokfirst(evsys_transtab, flags, &i); p;
		     p = trans_toknext(evsys_transtab, flags, &i))
			debug(1, ("%s: %s", dp->dirname, p));
	}
}


void
sie_init()
{
	int i;
	
	for (i = 0; i < sie_xlat[i].sie_mask; i++)
		defevt(trans_toktostr(sie_trans, sie_xlat[i].sie_mask),
		       &sie_xlat[i], 0);
}
	

int signo = 0;
int stop = 0;

void
sigmain(int sig)
{
	signo = sig;
	switch (signo) {
	case SIGCHLD:
	case SIGALRM:
		break;
	default:
		stop = 1;
	}
	signal(sig, sigmain);
}


#if USE_IFACE == IFACE_INOTIFY
# define INTERFACE "inotify"
# define INIT_EARLY 1
#elif USE_IFACE == IFACE_KQUEUE
# define INTERFACE "kqueue"
# ifdef HAVE_RFORK
#  define INIT_EARLY 1
# endif
#endif

static int opt_debug_level = 0;
static int opt_foreground = 0;
static char *opt_pidfile = NULL;
static char *opt_user = NULL;
static int opt_facility = -1;
static int lint_only = 0;

#include "cmdline.h"

int
main(int argc, char **argv)
{
	int i;
	struct grecs_node *tree;
	
	set_program_name(argv[0]);
	tag = (char*) program_name;

	sie_init();

	parse_options(argc, argv, &i);

	argc -= i;
	argv += i;

	switch (argc) {
	default:
		diag(LOG_CRIT, "too many arguments");
		exit(1);
	case 1:
		conffile = argv[0];
		break;
	case 0:
		break;
	}

	tree = grecs_parse(conffile);
	if (!tree)
		exit(1);

	config_finish(tree);
	if (lint_only)
		return 0;

	if (opt_debug_level)
		debug_level += opt_debug_level;
	if (opt_foreground)
		foreground = opt_foreground;
	if (opt_pidfile)
		pidfile = opt_pidfile;
	if (opt_facility != -1)
		facility = opt_facility;
	if (!foreground && facility <= 0)
		facility = LOG_DAEMON;
	if (opt_user)
		user = opt_user;
	
	if (facility > 0) {
		openlog(tag, LOG_PID, facility);
		grecs_log_to_stderr = 0;
	}

#ifdef INIT_EARLY
	setup_watchers();
#endif

	/* Become a daemon */
	if (!foreground) {
		if (daemon(0, 0)) {
			diag(LOG_CRIT, "daemon: %s", strerror(errno));
			exit(1);
		}
		log_to_stderr = -1;
	}
	

#ifndef INIT_EARLY
	setup_watchers();
#endif
	diag(LOG_INFO, "%s %s started", program_name, VERSION);

	/* Write pidfile */
	if (pidfile)
		storepid(pidfile);

	/* Relinquish superuser privileges */
	if (user && getuid() == 0)
		setuser(user);

	signal_setup(sigmain);

	/* Main loop */
	while (evsys_select() == 0 && !stop) {
		process_timeouts();
		process_cleanup(0);
	}

	diag(LOG_INFO, "%s %s stopped", program_name, VERSION);

	if (pidfile)
		unlink(pidfile);
	
	return 0;
}
